File: client/src/transport/compression.ts
import * as base64 from "base64-js";
import * as Crypto from "expo-crypto";
import * as pako from "pako";

import { SHARED_DICTIONARY } from "./dictionary";

export const ALGORITHM_GZIP = "gzip";
export const ALGORITHM_NONE = "none";
/** Raw deflate against a dictionary both ends hold. No header to amortise and a
 *  body of text to match against, so it wins where the other two cannot. */
export const ALGORITHM_DICT = "zd";

/** Truncated to match the server: 64 bits is plenty to catch mangled or
 *  mis-ordered chunks, and the full digest cost 48 bytes per envelope. */
export const CHECKSUM_HEX_LEN = 16;

export function compressToBase64(plaintext: string): string {
  return base64.fromByteArray(pako.gzip(new TextEncoder().encode(plaintext)));
}

export function decompressFromBase64(encoded: string): string {
  return new TextDecoder().decode(pako.ungzip(base64.toByteArray(encoded)));
}

export function compressWithDictionary(plaintext: string): string {
  return base64.fromByteArray(
    pako.deflateRaw(new TextEncoder().encode(plaintext), { dictionary: SHARED_DICTIONARY, level: 9 })
  );
}

export function decompressWithDictionary(encoded: string): string {
  return new TextDecoder().decode(
    pako.inflateRaw(base64.toByteArray(encoded), { dictionary: SHARED_DICTIONARY })
  );
}

/** Encoding must never make a payload bigger: compressing a short prompt adds a
 *  gzip header and then a third again for base64. Send whichever is smaller. */
export function encodePayload(plaintext: string): { algorithm: string; payload: string } {
  const candidates = [
    { algorithm: ALGORITHM_NONE, payload: plaintext },
    { algorithm: ALGORITHM_GZIP, payload: compressToBase64(plaintext) },
    { algorithm: ALGORITHM_DICT, payload: compressWithDictionary(plaintext) },
  ];
  return candidates.reduce((best, next) => (next.payload.length < best.payload.length ? next : best));
}

export function decodePayload(algorithm: string, payload: string): string {
  if (algorithm === ALGORITHM_GZIP) return decompressFromBase64(payload);
  if (algorithm === ALGORITHM_DICT) return decompressWithDictionary(payload);
  if (algorithm === ALGORITHM_NONE) return payload;
  throw new Error(`unknown algorithm: ${algorithm}`);
}

export async function sha256Hex(plaintext: string): Promise<string> {
  const digest = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, plaintext);
  return digest.slice(0, CHECKSUM_HEX_LEN);
}

File: server/app/protocol/compression.py
import base64
import gzip
import zlib
from typing import Tuple

from app.protocol.dictionary import SHARED_DICTIONARY

# Payload encodings, in the order they were added.
#
# "none" is the UTF-8 text as-is. "gzip" is base64(gzip(bytes)) — worth it for a
# long answer and actively harmful for a short one, because gzip adds a ~20 byte
# header and base64 then inflates the result by a third, so a small prompt could
# more than double.
#
# "zd" is raw deflate against a dictionary both ends already hold, base64'd. It
# has no header to amortise and starts with a body of text to match against, so
# it beats sending plainly even at ninety bytes — the crossover the other two
# leave in the middle simply is not there.
ALGORITHM_GZIP = "gzip"
ALGORITHM_NONE = "none"
ALGORITHM_DICT = "zd"

# Raw deflate: no zlib header, because the dictionary makes it unnecessary and
# the bytes are worth more than the formality.
_RAW_DEFLATE_WINDOW = -15


def compress_to_base64(data: bytes) -> str:
    return base64.b64encode(gzip.compress(data)).decode("ascii")


def decompress_from_base64(encoded: str) -> bytes:
    return gzip.decompress(base64.b64decode(encoded))


def compress_with_dictionary(data: bytes) -> str:
    compressor = zlib.compressobj(
        9, zlib.DEFLATED, _RAW_DEFLATE_WINDOW, 9, zlib.Z_DEFAULT_STRATEGY, zdict=SHARED_DICTIONARY
    )
    return base64.b64encode(compressor.compress(data) + compressor.flush()).decode("ascii")


def decompress_with_dictionary(encoded: str) -> bytes:
    decompressor = zlib.decompressobj(_RAW_DEFLATE_WINDOW, zdict=SHARED_DICTIONARY)
    return decompressor.decompress(base64.b64decode(encoded)) + decompressor.flush()


def encode_payload(data: bytes, dictionary: bool = False) -> Tuple[str, str]:
    """Return (algorithm, text) using whichever encoding is actually smaller.

    The dictionary is only offered to a caller that used it first. A relay that
    answered in an encoding the app has never heard of would be worse than a
    slow one — the answer would arrive and be unreadable.
    """
    candidates = [(ALGORITHM_NONE, data.decode("utf-8")), (ALGORITHM_GZIP, compress_to_base64(data))]
    if dictionary:
        candidates.append((ALGORITHM_DICT, compress_with_dictionary(data)))
    algorithm, text = min(candidates, key=lambda pair: len(pair[1]))
    return algorithm, text


def decode_payload(algorithm: str, text: str) -> bytes:
    if algorithm == ALGORITHM_GZIP:
        return decompress_from_base64(text)
    if algorithm == ALGORITHM_DICT:
        return decompress_with_dictionary(text)
    if algorithm == ALGORITHM_NONE:
        return text.encode("utf-8")
    raise ValueError(f"unknown algorithm: {algorithm}")
